src/engine/renderer/tr_shader.cpp

```diff
@@ -5265,6 +5314,136 @@ static void CollapseStages()
 			}
 		}
 	}
+
+	PileUpStages();
+}
+
+static void SetStagesRenderers()
+{
+	for ( int s = 0; s < shader.numStages; s++ )
+	{
+		shaderStage_t *stage = &stages[ s ];
+
+		/* stageType_t::ST_REFRACTIONMAP and stageType_t::ST_DISPERSIONMAP
+		are not implemented yet. */
+		static stageTypeRendererMap_t stageTypeRenderer = {
+			{ stageType_t::ST_COLORMAP, &Render_generic },
+			{ stageType_t::ST_STYLELIGHTMAP, &Render_generic3D },
+			{ stageType_t::ST_STYLECOLORMAP, &Render_generic3D },
+			{ stageType_t::ST_LIGHTMAP, &Render_lightMapping },
+			{ stageType_t::ST_DIFFUSEMAP, &Render_lightMapping },
+			{ stageType_t::ST_COLLAPSE_COLORMAP, &Render_lightMapping },
+			{ stageType_t::ST_COLLAPSE_DIFFUSEMAP, &Render_lightMapping },
+			{ stageType_t::ST_REFLECTIONMAP, &Render_reflection_CB },
+			{ stageType_t::ST_COLLAPSE_REFLECTIONMAP, &Render_reflection_CB },
+			{ stageType_t::ST_SKYBOXMAP, &Render_NOP },
+			{ stageType_t::ST_SCREENMAP, &Render_NOP },
+			{ stageType_t::ST_SKYBOXMAP, &Render_skybox },
+			{ stageType_t::ST_SCREENMAP, &Render_screen },
+			{ stageType_t::ST_PORTALMAP, &Render_portal },
+			{ stageType_t::ST_HEATHAZEMAP, &Render_heatHaze },
+			{ stageType_t::ST_LIQUIDMAP, &Render_liquid },
+			{ stageType_t::ST_ATTENUATIONMAP_XY, &Render_NOP },
+			{ stageType_t::ST_ATTENUATIONMAP_Z, &Render_NOP },
+		};
```

Comment:
I have a C++ question for this kind of `unordered_map`. I've done some testing on `godblot.org` and  it looks like when I write such `unordered_map` within a function GCC always rebuild it on every function run, at least the generated code is 2 or3 time longer.

If such `unordered_map` is written outside of the function, the map is done only once.

I wonder if there is a way to tell the compiler “this will never change, please creates it once for all”.

I noticed the same happens with basic arrays, but adding `static` makes GCC defines it globally and once for all, but here `static` doesn't help with maps.

Is there a way to tell the compiler “this will never change, please creates it once for all”?

I like the ability to set this in the block it is used instead of doing it globally, to not have to scroll the code that much in editor.

Reply:
`Ishq answered on IRC.